src/main/java/com/example/weatherwaveapi/service/impl/OpenWeatherServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+package com.example.weatherwaveapi.service.impl;
+
+import com.example.weatherwaveapi.model.request.WeatherRequest;
+import com.example.weatherwaveapi.model.response.*;
+import com.example.weatherwaveapi.model.response.openweather.WeatherOpenApiContainer;
+import com.example.weatherwaveapi.service.WeatherService;
+import com.example.weatherwaveapi.util.urlbuilder.OpenWeatherUrlBuilder;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.springframework.http.HttpMethod.GET;
+
+@RequiredArgsConstructor
+@Service
+@Slf4j
+public class OpenWeatherServiceImpl implements WeatherService {
+    private static final String EXCEPTION_MESSAGE = "HTTP error occurred while processing request. Exception message: {}";
+    private static final String ERROR_MESSAGE = "Failed to retrieve weather data ";
+    private static final String STRING_MESSAGE_FORMAT = "Search string: %s and error: %s";
+
+    private final RestTemplate restTemplate;
+    private final OpenWeatherUrlBuilder urlBuilder;
+
+    @Override
+    public WeatherResponse getWeather(WeatherRequest request) {
+        List<WeatherApiResponse> weatherResponses = new ArrayList<>();
+
+        if (request.cities() != null && !request.cities().isEmpty()) {
```

Comment:
Instead of this entry it is better to use: !CollectionUtils.isEmpty(request.cities())